File: java-pi-pantry/src/test/java/com/langrsoft/pi/pantry/APantry.java
package com.langrsoft.pi.pantry;

import static com.langrsoft.util.LambdaUtil.map;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static org.hamcrest.Matchers.*;
import static org.junit.Assert.*;
import java.util.List;
import org.junit.Test;

public class APantry {
   private Pantry pantry = new Pantry();

   @Test
   public void containsPurchasedItem() {
      pantry.purchase(new Item("cheerios"));

      boolean contains = pantry.contains("cheerios");

      assertThat(contains, is(true));
   }

   @Test
   public void doesNotContainUnpurchasedItem() {
      assertThat(pantry.contains("sugar"), is(false));
   }

   @Test
   public void hasCountZeroForItemsNotPurchased() {
      assertThat(pantry.count("sugar"), is(equalTo(0)));
   }

   @Test
   public void answersCountOfItemsByName() {
      pantry.purchase(new Item("sugar"));
      pantry.purchase(new Item("cheerios"));
      pantry.purchase(new Item("cheerios"));

      assertThat(pantry.count("cheerios"), is(equalTo(2)));
   }

   @Test
   public void answersCountOfAllItems() {
      pantry.purchase(new Item("sugar"));
      pantry.purchase(new Item("sugar"));
      pantry.purchase(new Item("cheerios"));
      pantry.purchase(new Item("cheerios"));

      int count = pantry.count();

      assertThat(count, is(equalTo(4)));
   }

   @Test
   public void storesAllInformationForPurchasedItems() {
      pantry.purchase(new ItemBuilder("sugar").withDescription("refined sweetener").create());

      Item retrieved = pantry.getItemNamed("sugar");

      assertThat(retrieved.getDescription(), is(equalTo("refined sweetener")));
   }

   @Test
   public void answersNullForUnpurchasedItem() {
      assertThat(pantry.getItemNamed("did not purchase"), is(nullValue()));
   }

   @Test(expected = IllegalArgumentException.class)
   public void throwsOnPurchaseWithNullName() {
      pantry.purchase(null);
   }

   @Test
   public void retrievesItemsByName() {
      pantry.purchase(new ItemBuilder("coffee").withDescription("small").create());
      pantry.purchase(new ItemBuilder("coffee").withDescription("large").create());

      List<Item> items = pantry.getItemsNamed("coffee");

      assertThat(map(items, Item::getDescription), equalTo(asList("small", "large")));
   }

   @Test
   public void returnsEmptyListWhenNoItemPurchasedWithName() {
      assertThat(pantry.getItemsNamed("pizza"), is(emptyList()));
   }

   @Test
   public void throwsWhenPurchasingAboveCapacity() {
      for (int i = 0; i < Pantry.CAPACITY; i++)
         pantry.purchase(new Item());

      try {
         pantry.purchase(new Item());
         fail();
      } catch (RuntimeException expected) {
      }
   }
}

File: csharp-pi-pantry/csharp-pi-pantry-test/DomainTests/APantry.cs
using System;
using NUnit.Framework;
using Pipantry.Domain;
using System.Linq;
using Moq;
using Pipantry.Util;
using csharp_pi_pantry_test.DomainTests;

namespace Test.Pipantry.Domain
{
    [TestFixture]
    public class APantry
    {
        Pantry pantry;
        readonly DateTime Now = DateTime.Now;

        [SetUp]
        public void Create()
        {
            pantry = new Pantry();
        }

        class AtAnyTime : APantry
        {
            [Test]
            public void ContainsPurchasedItem()
            {
                pantry.Purchase(new Item("cheerios"));

                var containsItem = pantry.Contains("cheerios");

                Assert.That(containsItem, Is.True);
            }

            [Test]
            public void DoesNotContainUnpurchasedItem()
            {
                Assert.That(pantry.Contains("sugar"), Is.False);
            }

            [Test]
            public void HasCountZeroForItemsNotPurchased()
            {
                Assert.That(pantry.Count("sugar"), Is.Zero);
            }

            [Test]
            public void AnswersCountOfItemsByName()
            {
                pantry.Purchase(new Item("sugar"));
                pantry.Purchase(new Item("cheerios"));
                pantry.Purchase(new Item("cheerios"));

                var count = pantry.Count("cheerios");

                Assert.That(count, Is.EqualTo(2));
            }

            [Test]
            public void AnswersCountOfAllItems()
            {
                pantry.Purchase(new Item("sugar"));
                pantry.Purchase(new Item("sugar"));
                pantry.Purchase(new Item("cheerios"));
                pantry.Purchase(new Item("cheerios"));

                var count = pantry.Count();

                Assert.That(count, Is.EqualTo(4));
            }

            [Test]
            public void StoresAllInformationForPurchasedItems()
            {
                pantry.Purchase(new ItemBuilder("sugar").WithDescription("refined sweetener").Create());

                var retrieved = pantry.ItemNamed("sugar");

                Assert.That(retrieved.Description, Is.EqualTo("refined sweetener"));
            }

            [Test]
            public void ThrowsWhenPurchasingAboveCapacity()
            {
                for (var i = 0; i < Pantry.CAPACITY; i++)
                    pantry.Purchase(new Item(""));

                Assert.Throws<Exception>(() => pantry.Purchase(new Item("")));
            }

            [Test]
            public void AnswersNullForUnpurchasedItem()
            {
                Assert.That(pantry.ItemNamed("did not purchase"), Is.Null);
            }

            [Test]
            public void ThrowsOnPurchaseWithNullName()
            {
                Assert.Throws<ArgumentException>(() => pantry.Purchase(null));
            }

            [Test]
            public void RetrievesItemsByName()
            {
                pantry.Purchase(new ItemBuilder("coffee").WithDescription("small").Create());
                pantry.Purchase(new ItemBuilder("coffee").WithDescription("large").Create());

                var coffees = pantry.ItemsNamed("coffee");

                Assert.That(coffees.Select(item => item.Description),
                        Is.EquivalentTo(new string[] { "small", "large" }));
            }

            [Test]
            public void ReturnsEmptyListWhenNoItemsPurchasedWithName()
            {
                Assert.That(pantry.ItemsNamed("pizza"), Is.Empty);
            }
        }

        class RightNow: APantry
        {
            [SetUp]
            public void SetPantryClockToNow()
            {
                var mock = new Mock<IDateTime>();
                mock.Setup(dateTime => dateTime.Now).Returns(Now);
                pantry.DateTimeProvider = mock.Object;
            }

            [Test]
            public void StoresCurrentTimestampForPurchasedItems()
            {
                pantry.Purchase(new Item("eggs"));

                var retrieved = pantry.ItemNamed("eggs");

                Assert.That(retrieved.PurchaseDate, Is.EqualTo(Now));
            }

            [Test]
            public void AnswersItemsExpiringToday()
            {
                pantry.Purchase(new ItemBuilder("milk").WithExpirationDate(Now).Create());
                pantry.Purchase(new ItemBuilder("juice").WithExpirationDate(Now.AddDays(1)).Create());

                var items = pantry.ItemsExpiringToday();

                Assert.That(items.Select(item => item.Name),
                        Is.EqualTo(new string[] { "milk" }));
            }

            [Test]
            public void AnswersAllExpiredItems()
            {
                pantry.Purchase(new ItemBuilder("milk").WithExpirationDate(Now).Create());
                pantry.Purchase(new ItemBuilder("juice").WithExpirationDate(Now.AddDays(1)).Create());
                pantry.Purchase(new ItemBuilder("curdled milk").WithExpirationDate(Now.AddDays(-1)).Create());

                var items = pantry.AllExpiredItems();

                Assert.That(items.Select(item => item.Name),
                    Is.EquivalentTo(new string[] { "milk", "curdled milk" }));
            }
        }
    }
}
